test/mb_debug_optimus.py

```diff
@@ -0,0 +1,667 @@
+        self.num_fail += (flag & MASK_FAIL) >> 9
+        self.num_dupl += (flag & MASK_DUPL) >> 10
+        self.num_supl += (flag & MASK_SUPL) >> 11
+        
+    @staticmethod
+    def _combine_flags(flags, op='or', primary_only=False):
+        if primary_only:
+            try:
+                filtered_flags = [flag for flag in flags if flag & MASK_PRIMARY == 0]
+            except:
+                raise Exception("No primary alignment is available")
+        else:
+            filtered_flags = flags
+            
+        combined_flag = 0
+        if op == 'or':
+            for flag in filtered_flags:
+                combined_flag |= flag
+        elif op == 'xor':
+            for flag in filtered_flags:
+                combined_flag ^= flag
+        elif op == 'and':
+            for flag in filtered_flags:
+                combined_flag &= flag
+        else:
+            raise Exception(f'Unknown bitwise operation {op}')
+
+        return combined_flag
+    
+    def process_single(self, flags, op='or', primary_only=False):
+        """Takes an iterable of binary flags, combines them according to a given bitwise operation, and updates
+        the summary statistics."""
+        combined_flag = self._combine_flags(flags, op, primary_only)
+        self._add_flag_to_summary(combined_flag)
+
+    def process_difference(self, first_flags, second_flags, joint_op='xor', primary_only=False):
+        first_combined_flag = self._combine_flags(first_flags, op='or', primary_only=primary_only)
+        second_combined_flag = self._combine_flags(second_flags, op='or', primary_only=primary_only)
+        if joint_op == 'xor':
+            self._add_flag_to_summary(first_combined_flag ^ second_combined_flag)
+        elif joint_op == 'and':
+            self._add_flag_to_summary(first_combined_flag & second_combined_flag)
+        else:
+            raise Exception(f'Unknown joint operation {joint_op}')
+            
+    def print_summary(self):
+        print(f'num_multi_seg: {self.num_multi_seg}')
+        print(f'num_proper: {self.num_proper}')
+        print(f'num_umapped: {self.num_umapped}')
+        print(f'num_next_templ_unmapped: {self.num_next_templ_unmapped}')
+        print(f'num_rc: {self.num_rc}')
+        print(f'num_next_rc: {self.num_next_rc}')
+        print(f'num_first_seg: {self.num_first_seg}')
+        print(f'num_last_seg: {self.num_last_seg}')
+        print(f'num_sa: {self.num_sa}')
+        print(f'num_fail: {self.num_fail}')
+        print(f'num_dupl: {self.num_dupl}')
+        print(f'num_supl: {self.num_supl}')
+# EPY: END code
+
+# EPY: START code
+op_multiple_primary = 0
+cr_multiple_primary = 0
+
+for idx in discordant_flag_indices:
+    op_flags = [attr.flag for attr in op_master_rec_attr_list[idx]]
+    cr_flags = [attr.flag for attr in cr_master_rec_attr_list[idx]]
+    
+    # reads with more than one primary alignment (!)
+    # according to SAM specs, each read should have only one primary alignment
```

Comment:
I used to use `--outSAMprimaryFlag AllBestScore`. This was changed in the below commit. Given this change, I can't see why multiple primary alignments would be found in the data unless there are _duplicate reads_ (i.e. those with the same read names).  https://github.com/HumanCellAtlas/skylab/commit/e64c243c0b9c3180c0d086b0d5f0755ae6ad68ff#diff-c8197e3f1fec7111d0322013f2bd8707